### PR TITLE
Display problems and game page redesign

### PR DESCRIPTION
**Pull request by helenli8181**
### List of Changes:

- `problem.ts`: Made type `Problem` into an export, and changed `getProblems` from a default to named export.
- `Game.tsx`: I added interfaces, changed the `GamePage` function into a class, and added a constructor and a render function which retrieve the problem data from `problem.ts`. Styling is contained within this class as style components.
- I used FlexBox to arrange the problem and code editor panels. 

### Notes:

- For now, it just automatically displays the first problem in the database. Of course, later we will need to be able to specify which problem, probably through random selection.
- If there is an error with finding the problem to display the error will be printed in the console. In the future it should display an error message on the webpage.  

### Screencast and Images:

Screen Shot 2020-09-11 at 5 19 07 PM

### CODE REVIEW

File: frontend/src/views/Game.tsx
Comment (alankbi) [inline frontend/src/views/Game.tsx:10]: What's the difference between interface and type? If we use interface here should we be doing the same elsewhere, or vice versa?

File: frontend/src/views/Game.tsx
Comment (alankbi) [inline frontend/src/views/Game.tsx:38]: nit: Could you replace `<p>` with `Text` from `src/components/core/Text.tsx` and create a new component for `<h3>` along the lines of the other components there? That way these font sizes can be reused on other pages too.

File: frontend/src/views/Game.tsx
Comment (alankbi) [inline frontend/src/views/Game.tsx:22]: We may want to add some error validation to display an error component if something goes wrong, although that might come later since it'd apply to all the different game pages

File: frontend/src/views/styles.module.css
Comment (alankbi) [inline frontend/src/views/styles.module.css:1]: nit: Move this into `Game.tsx` using [styled components](https://styled-components.com/), similar to how it's done [here](https://github.com/rocketden/main/blob/master/frontend/src/components/core/Button.tsx).

File: frontend/src/views/styles.module.css
Comment (alankbi) [inline frontend/src/views/styles.module.css:8]: Note for future reference: these colors might move into the Theme file once we come back and properly design everything

**Review by alankbi — CHANGES_REQUESTED** [review]
Page design looks great to me! Before merging, I requested a couple of changes and also had two notes/questions:

* What happens if no problems exist in the database, and do we need to worry about that (eventually the way problems are fetched will be changed, but idk if we should add an error check for that now or have a separate story to implement error checks for all pages)? 
* When I drag the screen width to below 640px, the components keep shrinking even though there's a global style setting min-width to 640px. Not sure why that happens, but I think it might have to do with using `%` vs. `vw`.

File: frontend/src/views/styles.module.css
Comment (zpChris) [inline frontend/src/views/styles.module.css:28]: nit: Add endline to the end of the file.

This is also an **eslint** suggestion, though that suggestion may not pop up on `css` files.

Comment (zpChris) [inline frontend/src/views/styles.module.css:28]: We already have a `header.tsx` file; I think we should use that instead of creating a new custom Header.

File: frontend/src/views/Game.tsx
Comment (zpChris) [inline frontend/src/views/Game.tsx:12]: I'm curious on @alankbi's thoughts on this; I was under the impression that we were going to use functions, and incorporate the `state` variable similar to [this React tutorial](https://reactjs.org/docs/hooks-state.html). I'm assuming you changed it to a class when seeing [the "State and Lifecycle" React tutorial](https://reactjs.org/docs/state-and-lifecycle.html) -- this may be better I'm just not sure, do you guys know what the pros / cons are?

File: frontend/src/api/problem.ts
Comment (zpChris) [inline frontend/src/api/problem.ts:1]: Comment on the whole `problem.ts` file: I think the merge conflicts was likely because I changed this file name to `Problem.ts` instead of `problem.ts`. That matches with the file naming pattern -- can you change this file name to `Problem.ts`?

File: frontend/src/api/problem.ts
Comment (zpChris) [inline frontend/src/api/problem.ts:13]: Note: No action required, but I'd like to get input on this; I think we determined last time that the `Problem[]` type doesn't actually type-check the function correctly. If this is the case, should we try to fix it, or is it not a big deal?

File: frontend/src/views/styles.module.css
Comment (zpChris) [reply frontend/src/views/styles.module.css:1]: Thoughts on also changing the styling to use **flexbox**? A guide on CSS-tricks is [here](https://css-tricks.com/snippets/css/a-guide-to-flexbox/#flexbox-background); I think this would be easier to design with in the future. I'd also like to get @alankbi's thoughts on this.

File: frontend/src/views/Game.tsx
Comment (zpChris) [inline frontend/src/views/Game.tsx:37]: Note: A scrollbar appears next to the `Editor` when you add new lines, even when it doesn't go past the bottom of the page. Is this fixable?

File: frontend/src/views/styles.module.css
Comment (zpChris) [inline frontend/src/views/styles.module.css:7]: nit: Add `padding` to the problem panel?

**Review by zpChris — CHANGES_REQUESTED** [review]
I have a couple requested changes on design layout, primarily on the `styles.module.css` file; I also left a few comments bringing up topics with no required changes, just choices I think we should discuss (for both @helenli8181 and @alankbi).

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:10]: I used interfaces here because they are more ideal for defining the props & states required by React.Component. Generally types and interfaces in Typescript are becoming more similar to each other but there are specific cases where one is better than the other, i.e. types rather than interfaces should be used as arguments or return types of a function. I think it's fine to have both in our program.

**Comment by helenli8181** [comment]
> Page design looks great to me! Before merging, I requested a couple of changes and also had two notes/questions:
> 
> * What happens if no problems exist in the database, and do we need to worry about that (eventually the way problems are fetched will be changed, but idk if we should add an error check for that now or have a separate story to implement error checks for all pages)?
> * When I drag the screen width to below 640px, the components keep shrinking even though there's a global style setting min-width to 640px. Not sure why that happens, but I think it might have to do with using `%` vs. `vw`.

@alankbi On the second point, I tried switching to vw instead of % but the window behaved the same way. In both cases the components continued to shrink as long as the browser window could.

File: frontend/src/api/problem.ts
Comment (helenli8181) [reply frontend/src/api/problem.ts:13]: I think it's fine to leave as is

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:57]: It looks like there's no way to make it disappear completely even while turning its visibility to "hidden". There doesn't seem to be an option to remove the large chunk of white space under the last line either.

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:42]: I'll add this to the trello board!

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:32]: The main difference between `useState` in a function and `setState` in a class is that `setState` allows you to define multiple state variables at once while `useState` can only have one. This purpose isn't really expressed in my code as it is because I'm using only one variable which specifies whether problems exist to be displayed. The `componentDidMount()` function in a class can be incredibly useful -- it allows you to wait until other components are rendered before running a certain piece of code. I think I can transition this `GamePage` class into a function if you guys think it's better, I can probably even get rid of `componentDidMount()`,  however I don't think we should necessarily try to avoid using classes for the sake of consistency.

Comment (alankbi) [reply frontend/src/views/Game.tsx:28]: Yeah so I think React Hooks (the functional components + useXXX functions) are newer and the ["future"](https://reactjs.org/docs/hooks-faq.html#should-i-use-hooks-classes-or-a-mix-of-both) of React, but were only released recently and class components are still more common in existing apps/tutorials. For the sake of consistency and long-term prospects, it might be good to stick with functional components, but I think the difference isn't too important and we can always convert them later.

File: frontend/src/api/problem.ts
Comment (alankbi) [reply frontend/src/api/problem.ts:13]: To update on this, I think [this link](https://stackoverflow.com/questions/55960602/how-to-cast-fetch-api-response-to-a-specific-type-in-typescript) explains the behavior well; basically, TypeScript can't actually validate the API responses match the defined types (cause it comes in at runtime), so it's just up to us to make sure it matches properly (so that at compile time in your IDE TypeScript can work properly when you use those types elsewhere in the app).

File: frontend/src/views/Game.tsx
Comment (alankbi) [reply frontend/src/views/Game.tsx]: Maybe that's part of the desired behavior? It seems like it makes it so that you can always scroll down just enough so that you don't see any of the lines you typed (IIRC VSCode does the same). Idk if there's a way to toggle that or not.

**Comment by alankbi** [comment]
> @alankbi On the second point, I tried switching to vw instead of % but the window behaved the same way. In both cases the components continued to shrink as long as the browser window could.

I think it's working now actually, once you decrease the width below 640px it stops shrinking so nice work on that :) 

I think most of my changes were addressed but I'll wait till @zpChris takes a look before doing doing a final run-through/approval if that sounds good

File: frontend/src/views/Game.tsx
Comment (zpChris) [reply frontend/src/views/Game.tsx:28]: I also want to add that I think `useState` allows you to set multiple state variables; I did so in PR #26. As for converting them later @alankbi, personally I don't think it would make sense to change it now just to change it back later -- whichever we choose, I think we should use now, as it'd make development much easier.

**Comment by zpChris** [comment]
[Display problems on game page](https://trello.com/c/oA6wA3WY/83-display-problems-on-game-page)

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:28]: Sounds good, I am switching the class into a function.

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx]: I'll make a note to look more into this issue when we're working more on the design! I think what's happening is there's an empty block under the code editor that stays the same size even when the editor expands.

**Comment by helenli8181** [comment]
New screenshot after merging with Master: 
Screen Shot 2020-09-14 at 2 09 20 PM

File: frontend/src/views/Game.tsx
Comment (zpChris) [inline frontend/src/views/Game.tsx:25]: I'm having trouble understanding if some of these attributes are actually contributing anything? Notably, the following don't seem to be doing anything:
```
flex: 1;
width 50vw;
```
Since you already have the `flex: 1` value, I don't know that the others are necessary?

I'd also consider looking into how you can update this to make the height of "Header + Problem & Code Panel" equal to the screen height (I also referenced this in a comment below).

File: frontend/src/views/Game.tsx
Comment (zpChris) [inline frontend/src/views/Game.tsx:32]: Add ending semicolon to the line.

Note: You should be seeing this (also ending newline) through ESLint.

File: frontend/src/views/Game.tsx
Comment (zpChris) [inline frontend/src/views/Game.tsx:28]: nit: Either keep this and change the other `views` files to this format, or change this to `function GamePage()` and keep the other files as-is. Personally, consistency of structure across files is a personal preference for me, and I think will help with readability moving forward.

File: frontend/src/views/Game.tsx
Comment (zpChris) [reply frontend/src/views/Game.tsx:42]: This is more of an open discussion, both for @alankbi and @helenli8181, as I'm open to either. I wanted to propose resolving this immediately within this PR, because I think it may be good to say that when you introduce `catch` functions that are _going to stay_ (this isn't temporary code, like in PR #23), if the PR is not already bloated we add it to that same PR -- what are both of your thoughts?

File: frontend/src/views/Game.tsx
Comment (zpChris) [inline frontend/src/views/Game.tsx:46]: Add error handling if there are no problems?

**Review by zpChris — CHANGES_REQUESTED** [review]
The design looks better, thank you! I'm still requesting changes because the code has some ESLint errors, you accidentally left the old merged code in some places which means the header is below the first code editor, and some of the styling seems superfluous.

File: frontend/src/views/Game.tsx
Comment (zpChris) [inline frontend/src/views/Game.tsx:14]: Forgot to add this during the review, so leaving it as a single comment -- could these Container components go in the `Container.tsx` `core` file?

File: frontend/src/components/core/Text.tsx
Comment (helenli8181) [reply frontend/src/components/core/Text.tsx:20]: Medium is a font size, but it's the default value so technically line 20 is not necessary. I kept it there just to have a `ProblemHeaderText` because we'll probably need more styling on it later.

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:16]: You're right, it doesn't do anything anymore. Thanks for pointing it out!

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:25]: Ok I got rid of those two lines, they may have been from an old commit that got mixed in

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:46]: The error handling is supposed to be represented by line 42. Here the question mark accounts for the possibility of problems not existing. When I test this with only `npm start` so that it can't access the problem data, it prints an error to the console. Other than changing the error validation as in the above comment, are there any other ways you'd recommend doing this?

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:45]: Sorry about that, I had rearranged the components but didn't notice that this version was not updated.

File: frontend/src/views/Game.tsx
Comment (zpChris) [reply frontend/src/views/Game.tsx:43]: I'd recommend displaying error text to show that an error occurred when trying to get something to load: see what @alankbi did in his PR [here](https://github.com/rocketden/main/pull/24/files#diff-c47bfa9c3f1b80608df1f7a1dc057e30R48), for example.

File: frontend/src/views/Game.tsx
Comment (zpChris) [reply frontend/src/views/Game.tsx:45]: What are your thoughts on styling it now?

**Comment by helenli8181** [comment]
Error message displayed when problem cannot load:

Screen Shot 2020-09-14 at 7 55 45 PM

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx]: The Editor only fills the space within its `FlexPanel` component; this works even with other FlexBox components occupying the page as shown by the new `RoomInfoBar` component.

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:49]: Just put in a proper error handling system; I tried to make it as consistent as possible with Alan's in `Landing.tsx` though there are a few small differences.

File: frontend/src/components/core/Text.tsx
Comment (zpChris) [reply frontend/src/components/core/Text.tsx:20]: Where do you see Medium is a font size? The below image shows the font sizes I can see -- where is Medium? If it is equal to `default`, why not just use `default` like we have elsewhere?
Medium Font Size Not Found

File: frontend/src/views/Game.tsx
Comment (zpChris) [reply frontend/src/views/Game.tsx]: Sorry, I think I didn't clarify -- IMO the page design should match [LeetCode](https://leetcode.com/problems/single-number-ii/), where you cannot scroll any further on the page (just within the divs, but the bottom of the page remains the footer with things like "Console" and "Submit"). I was asking if you could construct this so it would match that layout; essentially, so you cannot scroll on the whole page.

File: frontend/src/views/Game.tsx
Comment (zpChris) [inline frontend/src/views/Game.tsx:48]: nit: Reduce distance between this and the header? The below image seems like a big gap to me.
Info Bar and Header

File: frontend/src/views/Game.tsx
Comment (zpChris) [reply frontend/src/views/Game.tsx]: Bump @helenli8181 -- I think you forgot to respond to this before re-request?

File: frontend/src/views/Game.tsx
Comment (zpChris) [inline frontend/src/views/Game.tsx:30]: I believe this is superfluous, as it is already set above.

File: frontend/src/views/Game.tsx
Comment (zpChris) [inline frontend/src/views/Game.tsx:33]: Same here; not needed I believe.

**Review by zpChris — CHANGES_REQUESTED** [review]
Note: Most of my changes were left as comments -- not sure how this happened, sorry about that. You should be able to see more comments above, let me know if you have any trouble addressing them.

Thanks for making those changes! I'm requesting a few more edits, mainly because I want to try to get the whole game site to all be on one page without scrolling (I think that was miscommunicated earlier, sorry if that wasn't clear in my last review). Also, I have some nits on error handling.

I think I'm going to wait for @alankbi to review before I review again. 👍

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:30]: Resetting to null is safety for when this `useEffect()` function is called multiple times in one session. Originally, my `componentsDidMount()` equivalent would only have been called once but after merging with Alan's code it's supposed to rerun with every location change.

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:33]: Same response as above

File: frontend/src/components/core/Text.tsx
Comment (helenli8181) [reply frontend/src/components/core/Text.tsx:20]: Good point, I'll fix that.

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:52]: Got rid of the `<ErrorText>` tag; I didn't realize that the red style was a part of `<ErrorMessage>` already.

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:48]: This gap is mainly due to the Header being very tall while the words are squished at the top. Since the Header affects every page, do you think I should change it?

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx]: Oh I see what you mean. The page should fit the browser window now.

File: frontend/src/views/Game.tsx
Comment (zpChris) [reply frontend/src/views/Game.tsx:48]: I'll leave that largely up to your discretion -- if you do that, make sure the other pages look OK too. If not, I'd just modify the margin of the `FlexInfoBar` (you could also make this margin negative, but I'd be careful when doing that as I'm not sure if it's good design practice).

File: frontend/src/components/core/Container.tsx
Comment (alankbi) [inline frontend/src/components/core/Container.tsx:10]: Unrelated to your PR but I just realized that this code I wrote is probably completely unnecessary haha

File: frontend/src/components/core/Container.tsx
Comment (alankbi) [inline frontend/src/components/core/Container.tsx:32]: I'm actually not sure if these two should go in the container file; at the same time, I think it does make sense to have the flex container in this file, and these three components should be grouped together. So idk really what the best situation is here. Can FlexContainer be reused in other classes (I'm not too familiar with flex)? If so, then maybe keep that in here and over flex info bar/panel back to Game.tsx. If it's pretty specific to the game page/only is useful with the info bar and panel, then I think maybe it can actually be put back in the Game.tsx file (maybe renaming it to Content that way it doesn't get flagged for refactoring later on).

I'm not really familiar or sure on what the best route forward is with this, so leaving it for now should be also fine.

File: frontend/src/components/core/Text.tsx
Comment (alankbi) [reply frontend/src/components/core/Text.tsx:20]: Has this been fixed yet?

File: frontend/src/views/Game.tsx
Comment (alankbi) [reply frontend/src/views/Game.tsx:20]: Has this been fixed yet?

File: frontend/src/views/Game.tsx
Comment (alankbi) [reply frontend/src/views/Game.tsx:33]: Same as my previous comment about whether this has been changed to `setError('')`. I may also just be looking at an outdated version so sorry if that's the case

File: frontend/src/views/Game.tsx
Comment (alankbi) [reply frontend/src/views/Game.tsx:38]: As discussed in the call, I think we want to have some error check in case zero problems are returned (an empty list). As for how to go about that, not too sure, but a proposed solution would maybe be in the `getProblems` then block, add another else-if to see if `problems[]` is empty (if so call setError with a custom message), and if it's not, then just save the first problem in state (rather than the whole list).

File: frontend/src/views/Game.tsx
Comment (alankbi) [reply frontend/src/views/Game.tsx:48]: Not sure what's causing the space being so large (probably partly header but also maybe has something to do with padding/margin of flex container/info bar), but if you want to customize the header bar, I'd recommend creating a new, say, GameHeader file in `navigation`, which can have custom styling and text. I'm actually not sure anymore whether that should be put in the GameLayout file or whether it can be put directly in this file, because if we want to store the room id in the header (as in the mockup), then we'll need to pass that info as props to the header, which might be hard if it's in a separate file.

File: frontend/src/views/Game.tsx
Comment (alankbi) [reply frontend/src/views/Game.tsx:52]: Not sure if this has been updated already, but if not, this can simply be `{error ? <ErrorMessage message={error} /> : null}` (without the ErrorText wrapped around it).

File: frontend/src/views/Game.tsx
Comment (alankbi) [inline frontend/src/views/Game.tsx:55]: Not sure if the full-height thing has been fixed, but if not, I think it may have to do with the `100vh` here. Not sure if the solution would be to get rid of it (making it default back to `100%`) or something else though.

**Review by alankbi — CHANGES_REQUESTED** [review]
Mainly minor nits and suggestions/questions building on Chris' comments, but otherwise it's almost there I think!

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx]: Working on it now :)

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:38]: I just made a change where if the problem set is an empty array in `Game.tsx` then it displays a different error message. It doesn't seem like good practice to put if/else statements in the then block in getProblems(), if that's even possible.

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:57]: Yep, it got fixed --  it was actually changing % to vh that did it haha

File: frontend/src/views/Game.tsx
Comment (helenli8181) [reply frontend/src/views/Game.tsx:50]: I reduced the height of the room info bar; it looks slightly better but a header redesign is probably required later.

File: frontend/src/components/core/Container.tsx
Comment (alankbi) [inline frontend/src/components/core/Container.tsx:31]: Minor nit: Move color to Theme

File: frontend/src/views/Game.tsx
Comment (alankbi) [reply frontend/src/views/Game.tsx:40]: Sounds good; also, I deleted all problems in the database to see how `problems?.[0]` would behave and everything seems to be good 👍

**Review by alankbi — APPROVED** [review]
Thanks for addressing the changes! I left another minor nit to move one of the colors to the Theme file, but after that, looks good to me 👍

File: frontend/src/components/core/Container.tsx
Comment (zpChris) [reply frontend/src/components/core/Container.tsx:31]: Bump @helenli8181

File: frontend/src/views/Game.tsx
Comment (zpChris) [inline frontend/src/views/Game.tsx:45]: Thoughts on adding this `Header` to the `Game.tsx` page, similar to `Main.tsx`?

**Review by zpChris — APPROVED** [review]
LGTM! I want to note that there are changes requested before you merged, including one from @alankbi you still need to address. Please do those and make sure all comments are resolved before merging.

After that, let's merge! 🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉

File: frontend/src/components/core/Container.tsx
Comment (helenli8181) [reply frontend/src/components/core/Container.tsx]: I changed the game page to use the Theme background color so it matches our other pages now. The only drawback of this is the header and problem panel are now the same color.

**Pull request merged**
